test/rockpaperscissors.js

```diff
@@ -0,0 +1,772 @@
+const truffleAssert = require('truffle-assertions');
+const timeMachine = require('ganache-time-traveler');
+const Game = artifacts.require("./RockPaperScissors.sol");
+
+contract('rcp', async accounts => {
+
+    const { toBN, soliditySha3 } = web3.utils;
+    const { getBalance, getTransaction } = web3.eth;
+    const playerOneSecretString = "secret1";
+    const playerTwoSecretString = "secret2";
+    const wrongSecretString = "wrongSecret";
+    
+    const ROCK = toBN(1);
+    const PAPER = toBN(2);
+    const SCISSORS = toBN(3);
+
+    const DRAW = "0";
+    const LEFTWIN = "1";
+    const RIGHTWIN = "2";
+    const INCORRECT = "3";
+
+    const INIT = "0";
+    const PLAYER_ONE_MOVE = "1";
+    const PLAYER_TWO_MOVE = "2";
+    const PLAYER_ONE_REVEAL = "3";
+
+    const INVALID_MOVE_MSG = "Invalid move";
+    const INVALID_STEP_MSG = "Invalid step";
+    const HASH_MISMATCH_MSG = "Move and secret do not match";
+    const GAME_NOT_EXPIRED_MSG = "Game has not expired";
+    const MOVE_HASH_EMPTY_MSG = "Move hash is empty";
+    const SECRET_EMPTY_MSG = "Secret is empty";
+    const INVALID_PLAYER_MSG = "Invalid player";
+
+    const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
+    const ZERO_BYTES_32 = "0x0000000000000000000000000000000000000000000000000000000000000000";
+    const DAY_24_HOUR_IN_SECS = 86400;
+    const TIMEOUT = 3500000
+
+    const getBlockTimeStamp = async(txObj) => {
+
+        const tx = await getTransaction(txObj.tx);
+        const blockData = await web3.eth.getBlock(tx.blockNumber)
+
+        return blockData.timestamp;
+    }
+
+    const getGasCost = async txObj => {
+        const tx = await getTransaction(txObj.tx);
+
+        return toBN(txObj.receipt.gasUsed).mul(toBN(tx.gasPrice));
+    };
+
+    const [contractOwner, playerOne, playerTwo, playerThree] = accounts;
+    const playerOneSecretBytes32 = await soliditySha3(playerOneSecretString);
+    const playerTwoSecretBytes32 = await soliditySha3(playerTwoSecretString);
+    const wrongSecretBytes32 = await soliditySha3(wrongSecretString);
+
+    let rcp, gameKey, playerTwoMoveHash, snapshotId;
+
+    beforeEach("Deploy and prepare", async function() {
+        rcp = await Game.new({from: contractOwner});
+        gameKey = await rcp.createPlayerOneMoveHash(playerOne, playerOneSecretBytes32, ROCK);
+        playerTwoMoveHash = await rcp.createPlayerTwoMoveHash(playerTwo, gameKey, playerTwoSecretBytes32, PAPER);
+        const snapShot = await timeMachine.takeSnapshot();
+        snapshotId = snapShot['result'];
+    });
+
+    afterEach(async() => {
+        await timeMachine.revertToSnapshot(snapshotId);
```

Comment:
I never tried it, but I am thinking. How about create the snapshot in a `before`, and revert in an `afterEach`? Like that you do not expend again the computing cost of the snapshot creation.

Reply:
I moved these into the describe() sections where used but didn't seem to make much difference.